ext/phar/phar_object.c

```diff
@@ -4299,21 +4299,54 @@ static int phar_extract_file(zend_bool overwrite, phar_entry_info *entry, char *
 }
 /* }}} */
 
+static int extract_helper(phar_archive_data *archive, char *search, char *pathto, size_t pathto_len, zend_bool overwrite, char **error) { /* {{{ */
+	int extracted = 0;
+	phar_entry_info *entry;
+	int len;
+
+	/* nothing to match, or matching "/" -- extract all files */
+	len = search ? strlen(search) : 0;
+	if (0 == len || 0 == strcmp((const char *)search, "/")) {
+		ZEND_HASH_FOREACH_PTR(&archive->manifest, entry) {
+			if (FAILURE == phar_extract_file(overwrite, entry, pathto, (int)pathto_len, error)) return -1;
+			extracted++;
+		} ZEND_HASH_FOREACH_END();
+
+	/* ends in "/" -- extract all entries having that prefix */
+	} else if ('/' == search[len - 1]) {
+		ZEND_HASH_FOREACH_PTR(&archive->manifest, entry) {
+			if (0 != strncmp(search, entry->filename, len)) continue;
+			if (FAILURE == phar_extract_file(overwrite, entry, pathto, (int)pathto_len, error)) return -1;
+			extracted++;
+		} ZEND_HASH_FOREACH_END();
+
+	/* otherwise, looking for an exact match */
+	} else {
+		entry = zend_hash_str_find_ptr(&archive->manifest, search, len);
```

Comment:
The code previously used, depending on the case, either the stored length here, or directly used the zend_string for the lookup. I assume that phar paths can't contain null bytes, so binary safety is likely not an issue, but it would still be better to pass in a `zend_string` as `search` and then use `zend_hash_find_ptr` here.